src/storage/UpdateEdgeProcessor.cpp

```diff
@@ -0,0 +1,373 @@
+        }
+        VLOG(1) << "Hit srcProp for tag " << tagName << ", prop "
+                << prop << ", value " << it->second;
+        return it->second;
+    };
+    getters.getAliasProp = [&, this] (const std::string&,
+                                      const std::string &prop) -> OptVariantType {
+        auto res = RowReader::getPropByName(this->filterReader_.get(), prop);
+        if (!ok(res)) {
+            return Status::Error("Invalid Prop");
+        }
+        auto val = value(std::move(res));
+        VLOG(1) << "Hit edgeProp for edge prop " << prop << ", value " << val;
+        return val;
+    };
+    if (this->exp_ != nullptr) {
+        auto filterResult = this->exp_->eval();
+        if (filterResult.ok() && !Expression::asBool(filterResult.value())) {
+            VLOG(1) << "Filter skips the update";
+            return kvstore::ResultCode::ERR_FALSE_FILTER;
+        }
+    }
+
+    // prepare the return(yield clause) record
+    record_.resize(returnColumnsNum_);
+    if (returnColumnsNum_ > 0) {
+        std::string encoded = writer.encode();
+        auto writeSchema = std::make_shared<ResultSchemaProvider>(writer.moveSchema());
+        CHECK_EQ(returnColumnsNum_, writeSchema->getNumFields());
+        auto reader = RowReader::getRowReader(encoded, writeSchema);
+        int64_t index = 0;
+        for (auto& prop : this->edgeContext_.props_) {
+            if (prop.returned_) {
+                auto res = RowReader::getPropByIndex(reader.get(), index);
+                if (!ok(res)) {
+                    VLOG(3) << "Bad value for prop " << prop.prop_.get_name();
+                    return kvstore::ResultCode::ERR_FALSE_UPDATER;
+                }
+                record_[prop.retIndex_] = value(std::move(res));
+                index++;
+            }
+        }
+        for (auto& tc : tagContexts_) {
+            for (auto& prop : tc.props_) {
+                if (prop.returned_) {
+                    auto res = RowReader::getPropByIndex(reader.get(), index);
+                    if (!ok(res)) {
+                        VLOG(3) << "Bad value for prop " << prop.prop_.get_name();
+                        return kvstore::ResultCode::ERR_FALSE_UPDATER;
+                    }
+                    record_[prop.retIndex_] = value(std::move(res));
+                    index++;
+                }
+            }
+        }
+    }
+
+    // the Set clause for updating
+    SchemaVer ver = edgeReader_->schemaVer();
+    const auto constSchema = this->schemaMan_->getEdgeSchema(
+            spaceId_, this->edgeContext_.edgeType_, ver);
+    auto schema = std::const_pointer_cast<meta::SchemaProviderIf>(constSchema);
+    // TODO(zhangguoqing) When the edge is to be inserted, it can't instantiate
+    // the class RowUpdater() with edgeReader_ parameter, otherwise the updater_
+    // doing encoded will report the error: Failed to encode updated data.
+    if (resp_.get_upsert()) {
```

Comment:
I can't understand this code.   Why we check field in response?   And i think we don't need to support upsert in this processor,  we could do it in another one.

Reply:
I think, they are almost always the same, and it will very redundant when separated.

Reply:
It has been refactored.